### PR TITLE
added JWT auth, and some refactor

### PR DESCRIPTION
**Pull request by Oleksii-Melashchenko**

### CODE REVIEW

File: src/main/java/com/clozex/shop/controller/BookController.java [hunk 1]
```diff
@@ -36,14 +36,14 @@ public class BookController {
             - title
             - author
             """)
-    @PreAuthorize("hasRole('USER')")
+    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/clozex/shop/controller/BookController.java]: every user has a USER role. so you can validate only the USER role

File: src/main/java/com/clozex/shop/repository/role/RoleRepository.java [hunk 2]
```diff
@@ -2,9 +2,9 @@
 
 import com.clozex.shop.model.Role;
 import com.clozex.shop.model.RoleName;
-import java.util.Optional;
+import java.util.Set;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface RoleRepository extends JpaRepository<Role, Long> {
-    Optional<Role> findByName(RoleName roleName);
+    Set<Role> findByName(RoleName roleName);
```
Comment (liliia-ponomarenko) [inline src/main/java/com/clozex/shop/repository/role/RoleRepository.java]: Why Set?

File: src/main/java/com/clozex/shop/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -31,7 +30,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         }
         User user = userMapper.toModel(requestDto);
         user.setPassword(passwordEncoder.encode(user.getPassword()));
-        user.setRoles(Set.of(roleService.findByName(RoleName.ROLE_USER)));
+        user.setRoles(roleRepository.findByName(RoleName.ROLE_USER));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/clozex/shop/service/impl/UserServiceImpl.java]: return the previous realization

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes ;)

File: pom.xml [hunk 4]
```diff
@@ -95,6 +91,30 @@
             <artifactId>liquibase-maven-plugin</artifactId>
             <version>${liquibase.version}</version>
         </dependency>
+
+        <!-- Security -->
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-api</artifactId>
+            <version>0.12.6</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: create property for version

File: src/main/java/com/clozex/shop/config/OpenApiConfig.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class OpenApiConfig {
+
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+        .components(new Components().addSecuritySchemes("BearerAuth",
```
Comment (Elena-Bruyako) [inline src/main/java/com/clozex/shop/config/OpenApiConfig.java]: сheck tabulation
Create constant for all strings

File: src/main/java/com/clozex/shop/config/SecurityConfig.java [hunk 6]
```diff
@@ -42,7 +49,23 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
...
                         session ->
                                 session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
+                .addFilterBefore(jwtAuthenticationFilter,
+                        UsernamePasswordAuthenticationFilter.class)
                 .userDetailsService(userDetailsService)
                 .build();
     }
+
+    @Bean
+    AuthenticationManager authenticationManager(
```
Comment (Elena-Bruyako) [inline src/main/java/com/clozex/shop/config/SecurityConfig.java]:
```suggestion
    public AuthenticationManager authenticationManager(
```

File: src/main/java/com/clozex/shop/config/SecurityConfig.java [hunk 7]
```diff
@@ -42,7 +49,23 @@ public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Excepti
...
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
+        return authenticationConfiguration.getAuthenticationManager();
+    }
+
+    @Bean
+    public RoleHierarchy roleHierarchy() {
```
Comment (Elena-Bruyako) [inline src/main/java/com/clozex/shop/config/SecurityConfig.java]: why do you need this?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**